### DISCUSSION TITLE
apis 폴더 변경 건의

### DISCUSSION THREAD

**Discussion by sumi-0011** [git-goods/git-animal-client#34] [Ideas]
기존에는 나랩에서 했던 것과 같이 한 api당 한 파일 그리고 파일 이름을 use~ 로 시작하게 작명해두었습니다. 

→ 한 파일에 해당 API에 관련된 query와 request, response type만이 존재하므로 확인하기 편했던 경험이 있었기 때문

그런데, 나랩이랑 달리 API가 다양하고, 한 도메인에 관해 get, put, delete 요청들이 다른 파일에 위치하게되다보니 복잡함을 느꼈습니다.  
(ex, product 리스트 가져오기, 수정하기, 삭제하기 등)

→ 한 도메인에 대해서는 모두 한 파일에 작성하는건 어떨까? 라는 생각이 들어 아래와 같이 apis 구조를 변경하면 어떨까 하는 생각이 들었습니다. 
아래와 같이 바꾸는건 어떨까요?  @hyesungoh 

### 기존

```
📦apis
 ┣ 📂auctions
 ┃ ┣ 📜useGetHistory.ts
 ┃ ┣ 📜useGetMyProduct.ts
 ┃ ┣ 📜useGetProducts.ts
 ┃ ┣ 📜useProduct.ts
 ┃ ┗ 📜useRegisterProduct.ts
 ┣ 📂gotcha
 ┃ ┗ 📜useGotcha.ts
 ┣ 📂persona
 ┃ ┗ 📜useChangePersonaVisible.ts
 ┣ 📂user
 ┃ ┣ 📜getUsedCoupons.ts
 ┃ ┣ 📜useGetAllPets.ts
 ┃ ┣ 📜useGetUser.ts
 ┃ ┗ 📜useUseCoupons.ts
 ┣ 📜QueryClientProvider.tsx
 ┣ 📜index.ts
 ┣ 📜interceptor.ts
 ┗ 📜render.ts
```

### 변경후

```
📦apis
 ┣ 📂auctions
 ┃ ┣ 📜history.ts
 ┃ ┣ 📜my-product.ts
 ┃ ┣ 📜product.ts
 ┣ 📂gotcha
 ┃ ┗ 📜gotcha.ts
 ┣ 📂persona
 ┃ ┗ 📜persona.ts
 ┣ 📂user
 ┃ ┣ 📜coupons.ts
 ┃ ┣ 📜pet.ts
 ┃ ┣ 📜user.ts
 ┣ 📜QueryClientProvider.tsx
 ┣ 📜index.ts
 ┣ 📜interceptor.ts
 ┗ 📜render.ts
```

**Comment by hyesungoh** [comment]
흠 !! 일단 저는 레이어를 구분하면 좋겠다는 생각이 우선적으로 들긴 합니다

api를 직접 호출하는 레이어를 `apis`에 두고

react-query를 사용하는 레이어를 `hooks/api` 정도에 두는 느낌이랄까나요

```
api
ㄴ product
    ㄴ getProduct.ts
    ㄴ postProduct.ts

hooks
ㄴ api
    ㄴ product
        ㄴ useGetProduct.ts
        ㄴ usePostProduct.ts
```

약간 이런너낌??
---

`한 파일에 해당 도메인의 요청을 모아두자`

이 의견은 분류 기준이 명확해야될 것 같다는 생각이 들어요

```
apis
ㄴ 이곳의 폴더명의 기준과
    ㄴ 이 하위 파일명의 기준
```

한 파일에 모아두려면 기준이 명확해야 파악하기 쉬울 것 같은데, 

생각해두시거나 제가 알지 못하는 사항이 있으면 공유해주셔도 좋을거 같아용 👍

**Reply by sumi-0011** [reply]
우선적으로 하고싶었던 것은 `한 파일에 해당 도메인의 요청을 모아두자`인데, 분류기준은 나름..?! 명확했습니다. 
url이 같고, http method만이 다른 것들을 한 파일에 모아두고싶었던 것이였어요. 

저희 프로젝트에서 예시를 들자면, 
`https://api.gitanimals.org/auctions/products` url를 사용하고 있는 API들입니다. 
- https://github.com/git-goods/gitanimals-api/blob/main/docs/api/auction/change_product.md
- https://github.com/git-goods/gitanimals-api/blob/main/docs/api/auction/get_product.md
- https://github.com/git-goods/gitanimals-api/blob/main/docs/api/auction/change_product.md
- https://github.com/git-goods/gitanimals-api/blob/main/docs/api/auction/register_product.md

그런데 많긴..하네요..
번외로 뒤에 personaId만을 추가로 받는 [이런거](https://github.com/git-goods/gitanimals-api/blob/main/docs/api/auction/delete_product.md)도 있습니다

이런걸 한번에 모아두려면 파일 이름이 use~로 시작하게 되진 않게될거같아서 어떻게 분리하고 이름을 지을지 고민이 되네요..
